### PR TITLE
Add chapter about localization in PHP

### PR DESCRIPTION
**Pull request by linawolf**
references https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-CoreApi/issues/2286

releases: main, 11.5

### CODE REVIEW

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:10]:
```suggestion
Sometimes you have to localize a string in PHP code, for
```

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:27]: I find the usage of `camelCaseToLowerCaseUnderscored` overly complicated. In Extbase context I may have the extension name and can reuse it, mostly I have the extension key or just write it plain into the string.

If we go with this, we should link to the explanation of extension name/key, as many people don't know the difference, as I assume.

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:43]: Why static?

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:58]: I assume you mean the `translate()` method instead `pi_getLL()`?

**Review by brotkrueml — CHANGES_REQUESTED** [review]

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: $GLOBALS['LANG'] is mostly backend and cannot be used in frontend: When used in frontend and a backend user is logged in, the backend user language is used for lacalization.
See also: https://github.com/TYPO3/typo3/blob/main/typo3/sysext/core/Classes/Localization/LanguageService.php#L28

Mostly I use LocalizationUtility from Extbase. Don't know if this is best choice.

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (linawolf) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:43]: Because no Class context is needed.

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (linawolf) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: However $GLOBALS['LANG'] is used by the translation ViewHelper in the Core, which in turn is used in the frontend, so I figured that this is best practice?

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (linawolf) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:27]:
```suggestion
    $translationKey = 'LLL:EXT:my_extension'
```

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (linawolf) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:58]:
```suggestion
The function :php:`translate()` takes translation overrides from TypoScript into
```

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (linawolf) [inline Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:29]:
```suggestion
    $translationKey = 'LLL:EXT:my_extension' . '/Resources/Private/Language/locallang.xlf:' . $key;

```

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: Strange, I really had the problem an hour ago, that GLOBALS['LANG'] used the backend user language (custom "gsw_CH") for frontend ("de" and "en"). After switching to LocalizationUtility it worked correctly.

Comment (linawolf) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: Maybe it is only available in the View and therefore for ViewHelpers. The class comments states "Some Cirumstances"

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (linawolf) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst]: Removed it

File: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst
Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: And I am getting an error when not logged-in into backend: Call to a member function sL() on null. I just use a plain user funtion for testing. Maybe translate viewhelper initialises something before using?

Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: I remember vagely, that when the Admin Panel is activated, $GLOBALS['LANG'] works out of the box in frontend. That is obviously  then not the case without logged-in user.

Comment (linawolf) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: Here we also have the suggestion to use the Language Service: **https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Internationalization/XliffApi.html#label-access-in-php** However it does not state where we are at that moment :(

Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: "In the backend" -> this is correct. But not in frontend context.

Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: TSFE has a `sL` method which calls then ... tadaaa ... LanguageService:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L2513

And the language service is initialized in TSFE the LanguageServiceFactory:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L2524

This looks pretty good:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/core/Classes/Localization/LanguageServiceFactory.php#L46

But don't know if it is best practise to use that factory.

Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: Maybe just use the TSFE. But TSFE is not injectable anymore since v12 (it has a state). In user functions you can use the available content object renderer, in controllers it should be also available without global variable.

Comment (linawolf) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: Would you like to take over? For me it is starting to be to complicate

Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst:18]: I can make a draft for using TSFE in different contexts, but would like someone from core team to look over that before merging. Don't know if I can manage this in the next days. I'll try.

Then we can reference that chapter and suggest `TSFE->sL` for non-Extbase context and `LocalizationUtility` for Extbase context.

**Comment by linawolf** [comment]
@brotkrueml how shall we proceed with this PR? Should I maybe mak a separate PR describing the situation in Extbase first?

**Comment by brotkrueml** [comment]
Extbase is clear, go with `LocalizationUtility`.

In backend context use `$GLOBALS['LANG']`.

AbstractPlugin is outdated now, this should be adjusted, maybe here or in a follow-up.

And in plain PHP (frontend context, maybe also the non-Extbase plugin approach then): Core makes it now this way in the `TranslateViewHelper`:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/fluid/Classes/ViewHelpers/TranslateViewHelper.php#L213

So basically in frontend context:
```php
$languageServiceFactory = GeneralUtility::makeInstance(LanguageServiceFactory::class);
$languageService = $languageServiceFactory->createFromSiteLanguage($request->getAttribute('language')
    ?? $request->getAttribute('site')->getDefaultLanguage());
$languageService->sL(...)
```

Setting back `$GLOBALS['LANG']` in `TranslateViewHelper` is for caching, we should avoid this in a custom extension. If the language service is needed multiple times one can set a class property.

**Pull request closed without merging**
